src/github.com/matrix-org/dendrite/clientapi/routing/register.go

```diff
@@ -398,113 +196,38 @@ func Register(
-		sessions[sessionID] = append(sessions[sessionID], authtypes.LoginTypeSharedSecret)
-
-	case authtypes.LoginTypeApplicationService:
-		// Check Application Service register user request is valid.
-		// The application service's ID is returned if so.
-		appserviceID, err := validateApplicationService(cfg, req, r.Username)
-
-		if err != nil {
-			return *err
-		}
-
-		// If no error, application service was successfully validated.
-		// Don't need to worry about appending to registration stages as
-		// application service registration is entirely separate.
-		return completeRegistration(req.Context(), accountDB, deviceDB,
-			r.Username, "", appserviceID, r.InitialDisplayName)
-
-	case authtypes.LoginTypeDummy:
-		// there is nothing to do
-		// Add Dummy to the list of completed registration stages
-		sessions[sessionID] = append(sessions[sessionID], authtypes.LoginTypeDummy)
-
-	default:
-		return util.JSONResponse{
-			Code: 501,
-			JSON: jsonerror.Unknown("unknown/unimplemented auth type"),
-		}
-	}
-
-	// Check if the user's registration flow has been completed successfully
-	// A response with current registration flow and remaining available methods
-	// will be returned if a flow has not been successfully completed yet
-	return checkAndCompleteFlow(sessions[sessionID], req, r, sessionID, cfg, accountDB, deviceDB)
+	return HandleUserInteractiveFlow(req, r.userInteractiveFlowRequest, sessionID, cfg,
+		userInteractiveResponse{
+			// passing the list of allowed Flows and Params
+			cfg.Derived.Registration.Flows,
+			nil,
+			cfg.Derived.Registration.Params,
+			"",
+		}, wrapperFunction(accountDB, deviceDB, r))
 }
 
-// checkAndCompleteFlow checks if a given registration flow is completed given
-// a set of allowed flows. If so, registration is completed, otherwise a
-// response with
-func checkAndCompleteFlow(
-	flow []authtypes.LoginType,
-	req *http.Request,
-	r registerRequest,
-	sessionID string,
-	cfg *config.Dendrite,
-	accountDB *accounts.Database,
-	deviceDB *devices.Database,
-) util.JSONResponse {
-	if checkFlowCompleted(flow, cfg.Derived.Registration.Flows) {
-		// This flow was completed, registration can continue
-		return completeRegistration(req.Context(), accountDB, deviceDB,
-			r.Username, r.Password, "", r.InitialDisplayName)
-	}
-
-	// There are still more stages to complete.
-	// Return the flows and those that have been completed.
-	return util.JSONResponse{
-		Code: 401,
-		JSON: newUserInteractiveResponse(sessionID,
-			cfg.Derived.Registration.Flows, cfg.Derived.Registration.Params),
+// A function that will be called if flow completes
+// the call parameters of inside function are generic return parameters from flowhandler
+// the registration specific params are on wrapper function and are passed from handleRegistrationflow
+// while passing the wrapperFunction to handleUserInteractiveFlow
+func wrapperFunction(accountDB *accounts.Database, deviceDB *devices.Database, r registerRequest) func(
```

Comment:
Not sure how I feel about this function as a whole. Might be better to restructure where newUserInteractiveResponse just returns a `*JSONResponse` which may be `nil` depending on whether User Interactive Flow was completed or not. Once it does return nil, any function that's using User Interactive Flow can then continue forwards, assuming the user has passed authorization.

Reply:
Thats not so difficult....  if u dont want callbacks then I can return a 200 status reponse if flow succeeds and then check the response before calling final complete task function in the endpoint